File: fall_2022/Programming_languages/lab01/hxp4809_lab01.java
/*
 * Name: Hoang Anh Kiet Pham
 * ID: 1001904809
 * language version: jdk version 14.0.2
 * OS: Ubuntu (UTA virtual machine)
 */
import java.io.File;

public class hxp4809_lab01{
    
    public static void main(String[] args){
        long total = 0;
        String path_name = System.getProperty("user.dir"); 
        total = get_total_size_of_dir(path_name);
        System.out.println(total);
    }

    /*
     * this function is to get the total size of the current working directory
     * it takes the path of the directory as parameter,
     * then loop through every entry in that directory
     *     if the entry is a file:
     *         add its size to sum
     *     else:
     *         make the path to the child directory by simply add / and the directory name
     *         recursively call the function passing the path of child dir to it
     * return the sum
     */
    private static long get_total_size_of_dir(String path_name){
        File dir = new File(path_name);
        File[] entries = dir.listFiles();
        long sum = 0;
        for(File entry : entries){
            if(entry.isFile()){
                //System.out.println("file: " + entry.getName()); //for debug 
                sum += entry.length();
            }
            else{
                //System.out.println("dir: " + entry.getName()); //for debug
                String child_path_name = path_name + "/" + entry.getName();
                sum += get_total_size_of_dir(child_path_name);
            }
        }
        return sum;
     }

}

File: fall_2022/Programming_languages/lab01/hxp4809_lab01.py
'''
name: Hoang Anh Kiet Pham
Id: 1001904809
language version: python 3.8.10
OS: Ubuntu (UTA virtual machine)
'''

import os
'''
this function is to calculate the size of the current working directory
'''
def get_total_size_of_dir(dir):
    os.chdir(dir) #cd to sub directory
    entries = os.listdir() #get the list of all entry
    sum = 0
    for entry in entries:
        if os.path.isfile(entry):
            # print("file: " + entry) #for debug
            sum += os.path.getsize(entry)
        else:
            # print("dir: " + entry) #for debug
            sum += get_total_size_of_dir(entry) #recursively call this function
            os.chdir("..")
    return sum


#main function
total = get_total_size_of_dir(".")
print(total)

'''
extra question
1) was one language easier or faster to write the code for this?
if so, describe in detail why, as in what about the language made that the case
- for me, python and java aare easier and faster to work on this assignment
because they only require 1 library/class (File in java, os in python)
that provide every tools I need for this. Meanwhile, C is a little trickier.
It requires me to know more library (stat, file) and syntax than the others two.
Python is actually a little easier than java because it's a interpreted language
and very closed to human language.

2)Even though a langauge may not (e.g. FORTRAN) support recursion,
describe how you could write a program to produce the same results without using recursion.
Would that approach have any limitations and if so, what would they be?
- if the language don't support recursion, I would need to build a data structure (either stack or queue)
and process iteratively the file tree.
- This apprack would require us to write more complicated code
and need to handle the data passing around very carefully
- for some problem, figured the based case (conditon to stop the loop) is very difficult.
'''